src/wp-includes/html-api/class-wp-html-decoder.php

```diff
@@ -0,0 +1,457 @@
+<?php
+
+/**
+ * HTML API: WP_HTML_Decoder class
+ *
+ * Decodes spans of raw text found inside HTML content.
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since 6.6.0
+ */
+class WP_HTML_Decoder {
+	/**
+	 * Indicates if an attribute value starts with a given raw string value.
+	 *
+	 * Use this method to determine if an attribute value starts with a given string, regardless
+	 * of how it might be encoded in HTML. For instance, `http:` could be represented as `http:`
+	 * or as `http&colon;` or as `&#x68;ttp:` or as `h&#116;tp&colon;`, or in many other ways.
+	 *
+	 * Example:
+	 *
+	 *     $value = 'http&colon;//wordpress.org/';
+	 *     true   === WP_HTML_Decoder::attribute_starts_with( $value, 'http:', 'ascii-case-insensitive' );
+	 *     false  === WP_HTML_Decoder::attribute_starts_with( $value, 'https:', 'ascii-case-insensitive' );
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param string $haystack         String containing the raw non-decoded attribute value.
+	 * @param string $search_text      Does the attribute value start with this plain string.
+	 * @param string $case_sensitivity Optional. Pass 'ascii-case-insensitive' to ignore ASCII case when matching. Default 'case-sensitive'.
+	 * @return bool Whether the attribute value starts with the given string.
+	 */
+	public static function attribute_starts_with( $haystack, $search_text, $case_sensitivity = 'case-sensitive' ) {
+		$search_length = strlen( $search_text );
+		$loose_case    = 'ascii-case-insensitive' === $case_sensitivity;
+		$haystack_end  = strlen( $haystack );
+		$search_at     = 0;
+		$haystack_at   = 0;
+
+		while ( $search_at < $search_length && $haystack_at < $haystack_end ) {
+			$chars_match = $loose_case
+				? strtolower( $haystack[ $haystack_at ] ) === strtolower( $search_text[ $search_at ] )
```

Comment:
I mentioned this in the Token Map PR (https://github.com/WordPress/wordpress-develop/pull/5373#discussion_r1606722740) and now I've taken a moment to do some synthetic benchmarks, and I think we should change both that comparison in Token Map and this case to use `strcasecmp`. The difference is small, but may be relevant for highly optimized code. In ~3 million comparisons, `strcasecmp` ran 1.52 ± 0.02 times faster than `strtolower`.

```suggestion
				? 0 === strcasecmp( $haystack[ $haystack_at ], $search_text[ $search_at ] )
```

I also put these both into [3v4l.org](https://3v4l.org/6a2UE/vld) to get "Vulcan Logic Dumper" results, which does show `strcasecmp` performing fewer operations, supporting the idea that `strcasecmp` may perform slightly better:

```
Function str_case_cmp:
Finding entry points
Branch analysis from position: 0
1 jumps found. (Code = 62) Position 1 = -2
filename:       /in/6a2UE
function name:  str_case_cmp
number of ops:  9
compiled vars:  !0 = $a, !1 = $b
line      #* E I O op                           fetch          ext  return  operands
-------------------------------------------------------------------------------------
    3     0  E >   RECV                                             !0      
          1        RECV                                             !1      
    4     2        INIT_FCALL                                               'strcasecmp'
          3        SEND_VAR                                                 !0
          4        SEND_VAR                                                 !1
          5        DO_ICALL                                         $2      
          6        IS_IDENTICAL                                     ~3      $2, 0
          7      > RETURN                                                   ~3
    5     8*     > RETURN                                                   null

End of function str_case_cmp

Function str_to_lower:
Finding entry points
Branch analysis from position: 0
1 jumps found. (Code = 62) Position 1 = -2
filename:       /in/6a2UE
function name:  str_to_lower
number of ops:  11
compiled vars:  !0 = $a, !1 = $b
line      #* E I O op                           fetch          ext  return  operands
-------------------------------------------------------------------------------------
    6     0  E >   RECV                                             !0      
          1        RECV                                             !1      
    7     2        INIT_FCALL                                               'strtolower'
          3        SEND_VAR                                                 !0
          4        DO_ICALL                                         $2      
          5        INIT_FCALL                                               'strtolower'
          6        SEND_VAR                                                 !1
          7        DO_ICALL                                         $3      
          8        IS_IDENTICAL                                     ~4      $2, $3
          9      > RETURN                                                   ~4
    8    10*     > RETURN                                                   null
```

Reply:
We can land https://github.com/dmsnell/wordpress-develop/pull/15 to switch to `strcasecmp` here if you agree.

Reply:
please let me add another word of caution here, because you're comparing two very different functions. we don't know how long the comparison is that we want to make, so I suspect we'd need to perform multiple iterations through the short characters to find out. not a big deal, but with insignificant improvements, insignificant setbacks can add up too.

so the first question is _how many characters of the source text are involved in a match for a short word_? We don't know this up-front, which is why the existing algorithm moves character-by-character until it finds a result. that is, it proceeds until we have matches leading up to a null byte.

> I've taken a moment to do some **synthetic benchmarks**, and I think we should change both that comparison in Token Map and this case to use strcasecmp. The difference is small, but may be relevant for highly optimized code. In ~3 million comparisons, strcasecmp ran 1.52 ± 0.02 times faster than strtolower.

thanks for running these, but I'm not interested in synthetic benchmarks, particularly because of how misleading they can be. as with the broader measurements I've made in this patch, we're 6x - 30x slower than `html_entity_decode()` and yet the impact on WordPress is insignificant because we aren't spending our time on this.

in all of my explorations, both with 300,000 actual web pages and randomly-generated files full of character entities, _never_ were named character reference lookups in the hot path.

if we want to make this faster we should attack numeric character references, and I've failed to find any way to make that faster.

> strcasecmp performing fewer operations, supporting the idea that strcasecmp may perform slightly better:

I'll believe measurements and I'll use VLD dumps to gain insight, but not for performance. those operations can take wildly different amounts of time, plus two parallel `strtolower()` calls on independent data _might_ end up executing in parallel on the CPU, which can eliminate the sequential bottleneck. modern CPUs are complicated beasts.

---

I'm nervous about this because I find `strcasecmp()`'s interface much riskier because it's not doing what we want, but something different, and we have to adapt that different. so if you want to change this just make sure to include ample real-world metrics and try hard to break it, particularly with maps containing overlapping short tokens, e.g. `[ 'a' => '1', 'aa' => 2 ]`, and also maps with different prefix lengths.